go/tools/builders/nogo.go

```diff
@@ -0,0 +1,166 @@
+			haveCgo = true
+		} else {
+			goSrcs = append(goSrcs, src.filename)
+		}
+	}
+
+	workDir, cleanup, err := goenv.workDir()
+	if err != nil {
+		return err
+	}
+	defer cleanup()
+
+	imports, err := checkImports(srcs.goSrcs, deps, packageListPath, importPath, recompileInternalDeps)
+	if err != nil {
+		return err
+	}
+	cgoEnabled := os.Getenv("CGO_ENABLED") == "1"
+	if haveCgo && cgoEnabled {
+		// cgo generated code imports some extra packages.
+		imports["runtime/cgo"] = nil
+		imports["syscall"] = nil
+		imports["unsafe"] = nil
+	}
+	if coverMode != "" {
+		if coverMode == "atomic" {
+			imports["sync/atomic"] = nil
+		}
+		const coverdataPath = "github.com/bazelbuild/rules_go/go/tools/coverdata"
+		var coverdata *archive
+		for i := range deps {
+			if deps[i].importPath == coverdataPath {
+				coverdata = &deps[i]
+				break
+			}
+		}
+		if coverdata == nil {
+			return errors.New("coverage requested but coverdata dependency not provided")
+		}
+		imports[coverdataPath] = coverdata
+	}
+
+	importcfgPath, err := buildImportcfgFileForCompile(imports, goenv.installSuffix, filepath.Dir(outFactsPath))
+	if err != nil {
+		return err
+	}
+	if !goenv.shouldPreserveWorkDir {
+		defer os.Remove(importcfgPath)
+	}
+
+	return runNogo(workDir, nogoPath, goSrcs, facts, importPath, importcfgPath, outFactsPath, outLogPath)
+}
+
+func runNogo(workDir string, nogoPath string, srcs []string, facts []archive, packagePath, importcfgPath, outFactsPath string, outLogPath string) error {
+	if len(srcs) == 0 {
+		// emit_compilepkg expects a nogo facts file, even if it's empty.
+		return os.WriteFile(outFactsPath, nil, 0o666)
+	}
+	args := []string{nogoPath}
+	args = append(args, "-p", packagePath)
+	args = append(args, "-importcfg", importcfgPath)
+	for _, fact := range facts {
+		args = append(args, "-fact", fmt.Sprintf("%s=%s", fact.importPath, fact.file))
+	}
+	args = append(args, "-x", outFactsPath)
+	args = append(args, srcs...)
+
+	paramsFile := filepath.Join(workDir, "nogo.param")
+	if err := writeParamsFile(paramsFile, args[1:]); err != nil {
+		return fmt.Errorf("error writing nogo params file: %v", err)
+	}
+
+	cmd := exec.Command(args[0], "-param="+paramsFile)
+	out := &bytes.Buffer{}
+	cmd.Stdout, cmd.Stderr = out, out
+	// Always create this file as a Bazel-declared output, but keep it empty
+	// if nogo finds no issues.
+	outLog, err := os.Create(outLogPath)
+	if err != nil {
+		return fmt.Errorf("error creating nogo log file: %v", err)
+	}
+	defer outLog.Close()
+	if err := cmd.Run(); err != nil {
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			if !exitErr.Exited() {
+				cmdLine := strings.Join(args, " ")
+				return fmt.Errorf("nogo command '%s' exited unexpectedly: %s", cmdLine, exitErr.String())
+			}
+			prettyOut := relativizePaths(out.Bytes())
+			if exitErr.ExitCode() == 2 {
+				// Do not fail the action if nogo has findings so that facts are
+				// still available for downstream targets.
+				_, err := outLog.Write(prettyOut)
```

Comment:
```suggestion
			if exitErr.ExitCode() != 2 {
				return errors.New(string(prettyOut))
			}
		// Do not fail the action if nogo has findings so that facts are
		// still available for downstream targets.
		_, err := outLog.Write(prettyOut)
```
reduce nesting